ufl/core/external_operator.py

```diff
@@ -0,0 +1,278 @@
+# -*- coding: utf-8 -*-
+"""This module defines the ``ExternalOperator`` class, which symbolically represents operators that are not straightforwardly expressible in UFL. A practical implementation is required at a later stage to define how this operator should be evaluated as well as its derivatives from a given set of operands.
+"""
+
+# Copyright (C) 2019 Nacime Bouziani
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# Modified by Nacime Bouziani, 2020
+
+from ufl.coefficient import Coefficient
+from ufl.core.operator import Operator
+from ufl.core.ufl_type import ufl_type
+from ufl.constantvalue import as_ufl
+from ufl.log import error
+from ufl.finiteelement.finiteelement import FiniteElement
+from ufl.finiteelement.mixedelement import VectorElement, TensorElement
+from ufl.functionspace import FunctionSpace
+from ufl.domain import default_domain
+
+
+@ufl_type(num_ops="varying", inherit_indices_from_operand=0, is_differential=True)
+class ExternalOperator(Operator):
+
+    # Slots are disabled here because they cause trouble in PyDOLFIN
+    # multiple inheritance pattern:
+    _ufl_noslots_ = True
+
+    _external_operator_type = ''
+
+    def __init__(self, *operands, function_space, derivatives=None, count=None, coefficient=None, arguments=()):
+        r"""
+        :param operands: operands on which acts the :class:`ExternalOperator`.
+        :param function_space: the :class:`.FunctionSpace`,
+        or :class:`.MixedFunctionSpace` on which to build this :class:`Function`.
+        Alternatively, another :class:`Function` may be passed here and its function space
+        will be used to build this :class:`Function`.  In this case, the function values are copied.
```

Comment:
These lines need to be indented further for the RST to be correct.

Reply:
This looks like documentation from Firedrake (UFL has no `Function` object, nor does it have values).

Reply:
Fixed, thanks!